### PR TITLE
Allow job branch restrictions to be regex

### PR DESCRIPTION
**Pull request by stevekuznetsov**
When we have a branch `release-1.9` and want to quickly and quietly
create a derivative branch `release-1.9-hotfix-cve-1234`, we want the
derivative branch to have all of the same branch jobs as the branch it
is being derived from. In general, it should be cheap and easy to branch
out without having to modify too much of the testing infrastructure. One
step to making this possible is allowing branch restrictions on jobs to
be regular expressions. Git refs already have very stringent validation,
so we should not be concerned with trying to determine if a branch ref
as given is a regular expression or explicit branch name.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/area prow
/kind feature
/cc @smarterclayton @bentheelder
/assign @fejta @cjwagner @kargakis

The refactor was very annoying with the way we do config load -.-

### CODE REVIEW

File: prow/config/jobs.go [hunk 1]
```diff
@@ -309,21 +306,8 @@ func (c *Config) SetPresubmits(jobs map[string][]Presubmit) error {
 	for k, v := range jobs {
```
Comment (stevekuznetsov) [inline prow/config/jobs.go]: This is a super weird method...

File: prow/plugins/trigger/ic_test.go [hunk 2]
```diff
@@ -483,22 +499,26 @@ func TestHandleIssueComment(t *testing.T) {
...
+						AlwaysRun:    false,
+						Context:      "pull-jib",
+						Trigger:      `/test jib`,
+						RerunCommand: `/test jib`,
 					},
 				},
 			}
 		}
-		c.Config.SetPresubmits(presubmits)
+		if err := c.Config.SetPresubmits(presubmits); err != nil {
```
Comment (stevekuznetsov) [inline prow/plugins/trigger/ic_test.go]: Is there any lint or vet for ignored errors? Would have been useful here

File: prow/config/config.go [hunk 3]
```diff
@@ -898,9 +903,9 @@ func ValidateController(c *Controller) error {
 	return nil
 }
 
-// SetRegexes compiles and validates all the regural expressions for
+// SetPresubmitRegexes compiles and validates all the regural expressions for
```
Comment (cjwagner) [inline prow/config/config.go]: Typo.

File: prow/config/config.go [hunk 4]
```diff
@@ -920,6 +925,47 @@ func SetRegexes(js []Presubmit) error {
...
+		b, err := setBrancherRegexes(j.Brancher)
+		if err != nil {
+			return fmt.Errorf("could not set branch regexes for %s: %v", j.Name, err)
+		}
+		js[i].Brancher = b
+	}
+	return nil
+}
+
+// setBrancherRegexes compiles and validates all the regural expressions for
```
Comment (cjwagner) [inline prow/config/config.go]: Same typo.

File: prow/config/jobs.go (hunk 1 above)
Comment (cjwagner) [reply prow/config/jobs.go]: Agreed, its just for testing though.

**Comment by stevekuznetsov** [comment]
@cjwagner fixed typos in codebase

**Pull request merged**